quickwit-cluster/src/cluster.rs

```diff
@@ -0,0 +1,290 @@
+    let host_key;
+
+    if host_key_path.exists() {
+        let host_key_contents = fs::read(host_key_path)?;
+        host_key = Uuid::from_slice(host_key_contents.as_slice())?;
+        info!(host_key=?host_key, host_key_path=?host_key_path, "Read existing host key.");
+    } else {
+        if let Some(dir) = host_key_path.parent() {
+            if !dir.exists() {
+                fs::create_dir_all(dir)?;
+            }
+        }
+        host_key = Uuid::new_v4();
+        fs::write(host_key_path, host_key.as_bytes())?;
+        info!(host_key=?host_key, host_key_path=?host_key_path, "Create new host key.");
+    }
+
+    Ok(host_key)
+}
+
+/// A mamber information.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Member {
+    /// An ID that makes a member unique.
+    pub host_key: Uuid,
+
+    /// Listen address.
+    pub listen_addr: SocketAddr,
+
+    /// If true, it means self.
+    pub is_self: bool,
+}
+
+/// This is an implementation of a cluster using the SWIM protocol.
+pub struct Cluster {
+    /// A socket address that represents itself.
+    pub listen_addr: SocketAddr,
+
+    /// The actual cluster that implement the SWIM protocol.
+    artillery_cluster: Arc<ArtilleryCluster>,
+
+    /// A receiver(channel) for exchanging members in a cluster.
+    members: watch::Receiver<Vec<Member>>,
+}
+
+impl Cluster {
+    /// Create a cluster given a host key and a listen address.
+    /// When a cluster is created, the thread that monitors cluster events
+    /// will be started at the same time.
+    pub fn new(host_key: Uuid, listen_addr: SocketAddr) -> anyhow::Result<Self> {
+        info!( host_key=?host_key, listen_addr=?listen_addr, "Create new cluster.");
+        let config = ArtilleryClusterConfig {
+            cluster_key: CLUSTER_ID.as_bytes().to_vec(),
+            listen_addr,
+            ..Default::default()
+        };
+        let (artillery_cluster, _) =
+            ArtilleryCluster::new_cluster(host_key, config).map_err(|err| anyhow::anyhow!(err))?;
+
+        let (members_sender, members_receiver) = watch::channel(Vec::new());
+
+        // Create cluster.
+        let cluster = Cluster {
+            listen_addr,
+            artillery_cluster: Arc::new(artillery_cluster),
+            members: members_receiver,
+        };
+
+        // Add itself as the initial member of the cluster.
+        let member = Member {
+            host_key,
+            listen_addr,
+            is_self: true,
+        };
+        let initial_members: Vec<Member> = vec![member];
+        if members_sender.send(initial_members).is_err() {
+            error!("Failed to add itself as the initial member of the cluster.");
+        }
+
+        // Prepare to start a thread that will monitor cluster events.
+        let thread_inner = cluster.artillery_cluster.clone();
+        let thread_listen_addr = cluster.listen_addr;
+
+        // Start to monitor the cluster events.
+        tokio::task::spawn_blocking(move || {
```

Comment:
I wonder if we should actually really spawn a thread here (as opposed to using tokio::task::spawn_blocking